admin_terraform_version.go

```diff
@@ -0,0 +1,203 @@
+//
+// TFE API docs: https://www.terraform.io/docs/cloud/api/admin/terraform-versions.html
+type AdminTerraformVersions interface {
+	// List all the terraform versions.
+	List(ctx context.Context, options AdminTerraformVersionsListOptions) (*AdminTerraformVersionsList, error)
+
+	// Read a terraform version by its ID.
+	Read(ctx context.Context, id string) (*AdminTerraformVersion, error)
+
+	// Create a terraform version.
+	Create(ctx context.Context, options AdminTerraformVersionCreateOptions) (*AdminTerraformVersion, error)
+
+	// Update a terraform version.
+	Update(ctx context.Context, id string, options AdminTerraformVersionUpdateOptions) (*AdminTerraformVersion, error)
+
+	// Delete a terraform version
+	Delete(ctx context.Context, id string) error
+}
+
+// adminTerraformVersions implements AdminTerraformVersions.
+type adminTerraformVersions struct {
+	client *Client
+}
+
+// AdminTerraformVersion represents a Terraform Version
+type AdminTerraformVersion struct {
+	ID        string `jsonapi:"primary,terraform-versions"`
+	Version   string `jsonapi:"attr,version"`
+	URL       string `jsonapi:"attr,url"`
+	Sha       string `jsonapi:"attr,sha"`
+	Official  bool   `jsonapi:"attr,official"`
+	Enabled   bool   `jsonapi:"attr,enabled"`
+	Beta      bool   `jsonapi:"attr,beta"`
+	Usage     int    `jsonapi:"attr,usage"`
+	CreatedAt string `jsonapi:"attr,created-at"`
+}
+
+// AdminTerraformVersionsListOptions represents the options for listing
+// terraform versions.
+type AdminTerraformVersionsListOptions struct {
+	ListOptions
+}
+
+// AdminTerraformVersionsList represents a list of terraform versions.
+type AdminTerraformVersionsList struct {
+	*Pagination
+	Items []*AdminTerraformVersion
+}
+
+// List all the terraform versions.
+func (a *adminTerraformVersions) List(ctx context.Context, options AdminTerraformVersionsListOptions) (*AdminTerraformVersionsList, error) {
+	req, err := a.client.newRequest("GET", "admin/terraform-versions", &options)
+	if err != nil {
+		return nil, err
+	}
+
+	awl := &AdminTerraformVersionsList{}
+	err = a.client.do(ctx, req, awl)
+	if err != nil {
+		return nil, err
+	}
+
+	return awl, nil
+}
+
+// Read a terraform version by its ID.
+func (a *adminTerraformVersions) Read(ctx context.Context, id string) (*AdminTerraformVersion, error) {
+	if !validStringID(&id) {
+		return nil, ErrInvalidTerraformVersionID
+	}
+
+	u := fmt.Sprintf("admin/terraform-versions/%s", url.QueryEscape(id))
+	req, err := a.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	tfv := &AdminTerraformVersion{}
+	err = a.client.do(ctx, req, tfv)
+	if err != nil {
+		return nil, err
+	}
+
+	return tfv, nil
+}
+
+// AdminTerraformVersionCreateOptions for creating a terraform version.
+// https://www.terraform.io/docs/cloud/api/admin/terraform-versions.html#request-body
+type AdminTerraformVersionCreateOptions struct {
+	Type     *string `jsonapi:"primary,terraform-versions"`
```

Comment:
This is actually so awkward (in the existing API, I mean) that I'd go so far as to remove it entirely here in the client. The type here is a vestige of HashiCorp Atlas, when these 'tools' could be more than just Terraform.

As it's very purpose-built for Terraform now (it's even encoded in the route as `terraform-versions`), I think we should just remove this and provide it implicitly to the API (which I'm surprised doesn't just default to `terraform-versions`), not requiring the client user to have to think about a thing they have to provide and literally can't be anything _but_ `terraform-versions` 😆 

I'd either just document on the field that it's ignored and assign the required value in the functions when using the struct, or do the actual request with some sort of private struct that includes it (leaving it out of this public interface); whatcha think?

Reply:
Per the [conversation above](https://github.com/hashicorp/go-tfe/pull/186#discussion_r589703113), I think its better to keep these here, but not use a pointer, that way it is present as a default null value and that way its included in the `jsonapi` as part of marshalling.